SIP-19: Soft Bundle API (#19)
sui-foundation/sips: sips/sip-soft_bundle_api.md

|     Category | Core                                                 |
|      Created | 2024-04-05                                           |
| Comments-URI | https://sips.sui.io/comments-100                     |
|       Status | Draft                                                |
|     Requires | N/A                                                  |

## Abstract

This SIP introduces a Soft Bundle API to the Authority Server.

## Motivation

While PTB is a great way to bundle transactions, it is not suitable for sequencing transactions with different signers.
In addition, it does not allow partial reverts (e.g., a subset of the transactions in the PTB reverts). The Soft Bundle API is a way to bundle transactions with different signers and sequence them (and allow partial revert) in a single bundle with a high probability.

## Specification

Implement a new Grpc method `HandleSoftBundleCertificatesV2` in AuthorityServer, that accepts and executes a vector of certificates, ensuring:

- If at least one certificate cannot be executed, the whole request is denied.
- If at least one certificate has already been executed, the whole request is denied.

> - If at least one certificate contains only owned object, the whole request is denied.

Reviewer: ```suggestion
- If at least one certificate does not access a shared object, the whole request is denied.
```

- If at least one certificate has a different gas price than others, the whole request is denied.
- If the number of certificates exceeds `N` , the whole request is denied.
  N should be small enough, say `N = 4`.
- If the total number of BCS serialized bytes from all certificates exceeds M bytes, the whole request is denied. M should be small enough, say `M = 64K`.

The method submits certificates from the same bundle to consensus altogether, without delay.
Submission will be through the same Narwhal worker, in the same Batch that is broadcasted to all other workers on the network.
This requires modification to consensus client (ConsensusAdapter) as well as Narwhal’s BatchMaker, so that they accept a vector of certificates as input.

After including the bundle in the same Batch, we can be certain that if this Batch is getting included in a Header in consensus, its internal ordering will be respected.
Meanwhile:

> - Since all certificates may not contain only owned objects, their relative order is determined by consensus.

Reviewer: ```suggestion
- Since all certificates access at least one shared object, their relative order is determined by consensus.
```

- PostConsensusTxReorder does not affect this because all transaction blocks in the same bundle have identical gas prices.

The response type of the method will be:

```Rust
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct HandleSoftBundleCertificatesResponseV2 {
    pub responses: Vec<HandleCertificateResponseV2>,
}
```

> ## Rationality

Reviewer: ```suggestion
## Rationale
```
